File: agents/python/cowriter/agent.py
from __future__ import annotations

import asyncio
import json
from typing import Any

from livekit import agents
from livekit.agents import AutoSubscribe, JobContext, WorkerOptions, cli

from .models import CowriterModel, PrepPayload
from .prompt import build_system_prompt, greeting_for_mode
from .settings import get_settings
from .tools.music_gen import music_gen
from .tools.voice_swap import voice_swap


async def entrypoint(ctx: JobContext) -> None:
    """Primary LiveKit Agent coroutine."""

    settings = get_settings()
    await ctx.connect(auto_subscribe=AutoSubscribe.AUDIO_ONLY)

    # Listen for data-channel payloads (prep + further instructions)
    ctx.add_data_listener(lambda dc, p, d: asyncio.create_task(handle_data(ctx, d)))

    # If job metadata already contains prep info, handle it immediately
    if ctx.job_input:
        try:
            payload = json.loads(ctx.job_input)
            await handle_prep(ctx, payload)
        except Exception as exc:  # noqa: BLE001
            await ctx.send_data(json.dumps({"type": "agent/error", "error": str(exc)}).encode())

    # Keep running until room closed
    await ctx.wait_for_completion()


async def handle_data(ctx: JobContext, data: bytes) -> None:
    try:
        payload = json.loads(data.decode())
    except Exception as exc:  # noqa: BLE001
        await ctx.send_data(json.dumps({"type": "agent/error", "error": f"bad json: {exc}"}).encode())
        return

    if payload.get("type") == "prep":
        await handle_prep(ctx, payload)
    elif payload.get("type") == "user/text":
        await ctx.send_data(json.dumps({"type": "agent/text", "text": "Got it—what would you like me to do next?"}).encode())


async def handle_prep(ctx: JobContext, payload: dict[str, Any]) -> None:
    settings = get_settings()

    prep = parse_prep(payload)
    model = prep.model or CowriterModel(
        id="me-joel",
        first_name="Joel",
        bio="a singer/artist with a soulful voice and rich lyrics and melodies",
        voice_id=settings.elevenlabs_voice_id,
    )

    system_prompt = build_system_prompt(model)
    await ctx.send_data(json.dumps({"type": "agent/system", "prompt": system_prompt}).encode())
    await ctx.send_data(json.dumps({"type": "agent/text", "text": greeting_for_mode(prep.mode, model.first_name)}).encode())

    if prep.mode == "swap" and prep.vocal_url:
        await ctx.send_data(json.dumps({"type": "agent/text", "text": "Running voice swap…"}).encode())
        result = await voice_swap(prep.vocal_url, model.voice_id, settings.elevenlabs_api_key)
        await ctx.send_data(json.dumps(_tool_payload("voiceSwap", result)).encode())

    if prep.mode == "generate" and prep.prompt:
        await ctx.send_data(json.dumps({"type": "agent/text", "text": "Generating a musical idea…"}).encode())
        result = await music_gen(prep.prompt, prep.stem_url, settings.openai_api_key, settings.gemini_api_key)
        await ctx.send_data(json.dumps(_tool_payload("musicGen", result)).encode())


def parse_prep(payload: dict[str, Any]) -> PrepPayload:
    mode = payload.get("mode", "chat")
    model_info = payload.get("model")
    model = None
    if isinstance(model_info, dict):
        model = CowriterModel(
            id=model_info.get("id", "me-joel"),
            first_name=model_info.get("firstName", "Joel"),
            bio=model_info.get("bio", "a soulful singer"),
            voice_id=model_info.get("voiceId"),
        )
    return PrepPayload(
        type="prep",
        mode=mode,
        vocal_url=payload.get("vocalUrl"),
        prompt=payload.get("prompt"),
        stem_url=payload.get("stemUrl"),
        model=model,
    )


def _tool_payload(tool: str, result) -> dict[str, Any]:
    if result.ok:
        return {"type": f"tool/{tool}:ok", "url": result.artifact_url}
    return {"type": f"tool/{tool}:err", "error": result.error}


if __name__ == "__main__":
    cli.run_app(
        entrypoint,
        WorkerOptions(auto_subscribe=AutoSubscribe.AUDIO_ONLY),
    )

File: agents/cowriter/agent.ts
/*
  LiveKit Agent entrypoint (skeleton)
  - Greets based on pre-session context from client
  - Invokes voiceSwap/musicGen tools as needed
  - Streams back simple data messages with results

  NOTE: Replace stubbed tool calls with your production APIs.
*/

import 'cross-fetch/polyfill';
import WebSocket from 'ws';
import { buildSystemPrompt, greetingForMode } from './prompt';
import { musicGen } from './tools/musicGen';
import { voiceSwap } from './tools/voiceSwap';
import type { CowriterModel, PrepMessage, PrepMode } from './types';

// If you use @livekit/agents, import your Agent runner here.
// This file uses a lightweight Room WS skeleton to illustrate message flow without external deps.

interface Env {
  LIVEKIT_URL: string;
  LIVEKIT_API_KEY: string;
  LIVEKIT_API_SECRET: string;
  ELEVENLABS_API_KEY?: string;
  ELEVENLABS_VOICE_ID?: string;
  OPENAI_API_KEY?: string;
  GOOGLE_GEMINI_API_KEY?: string;
}

const env: Env = {
  LIVEKIT_URL: process.env.LIVEKIT_URL || '',
  LIVEKIT_API_KEY: process.env.LIVEKIT_API_KEY || '',
  LIVEKIT_API_SECRET: process.env.LIVEKIT_API_SECRET || '',
  ELEVENLABS_API_KEY: process.env.ELEVENLABS_API_KEY,
  ELEVENLABS_VOICE_ID: process.env.ELEVENLABS_VOICE_ID,
  OPENAI_API_KEY: process.env.OPENAI_API_KEY,
  GOOGLE_GEMINI_API_KEY: process.env.GOOGLE_GEMINI_API_KEY,
};

// Minimal logging utility
function log(...args: any[]) { console.log('[agent]', ...args); }

// This demonstrates message handling; replace with LiveKit Agents SDK runner when integrating.
async function startAgent() {
  if (!env.LIVEKIT_URL) {
    log('Missing LIVEKIT_URL');
    process.exit(1);
  }

  // In production, obtain a proper access token and connect via LiveKit Agents SDK.
  // For illustration, we use a placeholder WS connect to an arbitrary endpoint.
  const ws = new WebSocket(env.LIVEKIT_URL);

  ws.on('open', () => {
    log('Connected to LiveKit (WS placeholder)');
    // Announce readiness
    ws.send(JSON.stringify({ type: 'agent/ready' }));
  });

  ws.on('message', async (data) => {
    try {
      const msg = JSON.parse(String(data));
      if (msg?.type === 'prep') {
        const prep = msg as PrepMessage;
        await handlePrep(ws, prep);
      } else if (msg?.type === 'user/text') {
        // Extend: route to LLM to decide tool usage based on system prompt
        ws.send(JSON.stringify({ type: 'agent/text', text: 'Got it — what would you like me to do next?' }));
      }
    } catch (e) {
      log('invalid message', e);
    }
  });

  ws.on('close', () => log('WS closed'));
  ws.on('error', (e) => log('WS error', e));
}

async function handlePrep(ws: WebSocket, prep: PrepMessage) {
  const model: CowriterModel = prep.model || {
    id: 'me-joel',
    firstName: 'Joel',
    bio: 'a singer/artist with a soulful voice and rich lyrics and melodies',
    voiceId: process.env.ELEVENLABS_VOICE_ID,
  };

  const systemPrompt = buildSystemPrompt(model);
  ws.send(JSON.stringify({ type: 'agent/system', prompt: systemPrompt }));

  // Greet based on mode
  ws.send(JSON.stringify({ type: 'agent/text', text: greetingForMode(prep.mode as PrepMode, model.firstName) }));

  // Auto-run based on provided inputs
  if (prep.mode === 'swap' && prep.vocalUrl && model.voiceId) {
    ws.send(JSON.stringify({ type: 'agent/text', text: 'Running voiceSwap…' }));
    const res = await voiceSwap({ vocalUrl: prep.vocalUrl, targetVoiceId: model.voiceId });
    if (res.ok) {
      ws.send(JSON.stringify({ type: 'tool/voiceSwap:ok', url: res.artifactUrl }));
    } else {
      ws.send(JSON.stringify({ type: 'tool/voiceSwap:err', error: res.error }));
    }
  }

  if (prep.mode === 'generate' && prep.prompt) {
    ws.send(JSON.stringify({ type: 'agent/text', text: 'Composing preview with musicGen…' }));
    const res = await musicGen({ prompt: prep.prompt, stemUrl: prep.stemUrl });
    if (res.ok) {
      ws.send(JSON.stringify({ type: 'tool/musicGen:ok', url: res.artifactUrl }));
    } else {
      ws.send(JSON.stringify({ type: 'tool/musicGen:err', error: res.error }));
    }
  }
}

startAgent().catch((e) => {
  log('fatal', e);
  process.exit(1);
});

